subsys/net/ip/net_pkt.c

```diff
@@ -1654,6 +1654,59 @@ void net_pkt_get_info(struct k_mem_slab **rx,
 	}
 }
 
+int net_pkt_get_src_addr(struct net_pkt *pkt, struct sockaddr *addr,
+			 socklen_t *addrlen)
+{
+	struct net_context *ctx;
+	enum net_ip_protocol proto;
+	sa_family_t family;
+	u16_t port = 0;
+
+	if (!addr || !pkt || !addrlen) {
+		return -EINVAL;
+	}
+
+	ctx = pkt->context;
+	proto = net_context_get_ip_proto(ctx);
+	family = net_context_get_family(ctx);
+
+	addr->sa_family = family;
+
+	if (IS_ENABLED(CONFIG_NET_TCP) && proto == IPPROTO_TCP) {
+		struct net_tcp_hdr *hdr = net_pkt_tcp_data(pkt);
+
+		port = hdr->src_port;
+	} else if (IS_ENABLED(CONFIG_NET_UDP) && proto == IPPROTO_UDP) {
+		struct net_udp_hdr *hdr = net_pkt_udp_data(pkt);
+
+		port = hdr->src_port;
+	} else {
+		return -ENOTSUP;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
+		struct sockaddr_in6 *addr6 = net_sin6(addr);
+
+		if (*addrlen < sizeof(struct sockaddr_in6)) {
```

Comment:
What about returning the address len in *addrlen? If you didn't intend to do that, why is it passed by pointer?

Reply:
Yes, the address size isn't necessary to be a pointer.

Reply:
Hmm, I'd say "it's useful to return the actual addrlen". Because otherwise, we probably don't even need to pass it in - struct sockaddr can accommodate both IPv4 and IPv6 addresses.

Reply:
It's used to prevent out of bound access. For example, the address is IPv6 but a sockaddr_in variable is passed in. I know it's rare but the caller might make a mistake. In such case, an error will be returned, so I probably don't need to return the actual addrlen.

Reply:
Ok, I assume you considered various usecases for this function, so sounds good. (And we can change it later if found useful.)